Fix slow restart against the AzureJobStore (resolves #584)

adamnovak opened the pull request:
Basically, we now download all the jobs from the jobStore before trying
to "clean" the jobStore or rebuild the ToilState. This means we aren't
running individual per-job operations against every job in the jobStore,
which can be slow if the jobStore is Azure or AWS and each of those
operations requires a round-trip HTTP request.

To make this work, I also fixed the Azure job store's `jobs()` method,
which was only returning the first 1000 jobs.

The `clean()` method and the restart process in general is now much
noisier, with info messages letting you know how many jobs have been
loaded so far.

Fix logger naming in AzureJobStore

This still needs unit tests, but will close #584.

hannes-ucsc commented on src/toil/jobStores/azureJobStore.py:
```diff
@@ -120,8 +120,45 @@ def qualify(self, name):
+    
+        while True:
+            # Get a page (up to 1000 items)
+            page = self.jobItems.query_entities(
```
Can the paging functionality be placed in a reusable method of the AzureTable class?

adamnovak commented:
OK I added some tests. I'm not sure how to make a test log output (I tried print). It seems like running clean with the cache should always be faster than without the cache, so I've added that as an assert in the test of the clean function.

Still to do is making the paging reusable, and adding a test for toil state construction with the cache.

Also, in order to test the Azure jobs method yielding >1000 jobs, I need to actually make >1000 jobs. I do that in this, but it's going to be slow making so many trips to the job store and back.

hannes-ucsc commented:
That's ok. We'll probably skip this test for regular builds. But it would be great to expose the AWS store to the same test.

hannes-ucsc commented on src/toil/test/jobStores/jobStoreTest.py:
```diff
@@ -423,6 +424,58 @@ def assertUrl(self, url):
+            
+            cacheTime = cacheEnd - cacheStart
+            
+            print("Without cache: %s With cache: %s" % (str(noCacheTime), str(cacheTime)))
```
To use print as a function you would need to add `from __future__ import print_function` at the top.

adamnovak replied:
It seems perfectly happy with the syntax without that, though it might be parsed as a print statement and then a parenthesized string.

Do we use that elsewhere in Toil?

hannes-ucsc replied:
The syntax is valid but the parentheses are redundant and will be eliminated. It just looks as if you are using the print function. I would just remove the parentheses. I would only use the `print()` function if wanted access to its `file=` keyword argument. Choosing the output stream is the only feature that the `print` statement doesn't give you. I think we mix and match print statement and function in Toil, and I don't see a problem with that as long as we don't use visual ambiguity introduced by redundant parentheses.

adamnovak commented:
I'm not sure what's up with the tests. Jenkins failed it on `src/toil/test/sort/sortTest.py::SortTest::testFileSingle10000` with an error from the FileJobStore saying that a job that was supposed to exist didn't. Unfortunately, that test works just fine for me locally.

I've thrown out the print function; maybe that will magically make the tests pass.

adamnovak commented:
OK, I've successfully gotten it to fail. It looks like there are bad/corrupt jobs getting into the FileJobStore, and when I go through to load all of them to make the cache, the `.jobs()` iterator explodes because it can't deserialize some jobs.

It looks like either `.jobs()` needs to be modified to skip over jobs that can't be deserialized, or a custom override or version of `.clean()` has to somehow run first and clean up any damaged jobs.

hannes-ucsc commented:
Looks good. I think the correct procedure would be to merge this as is and open a new bug ticket against the FileJobStore. If you have a stack trace from the failure you mention in https://github.com/BD2KGenomics/toil/pull/585#issuecomment-160811287, can you post that here or add that to the new ticket.

For posterity, can you confirm that you saw `testManyJobs` pass at least once against `AzureJobStore`?

adamnovak commented:
The testManyJobs Azure test was (occasionally?) failing due to a conflict error on insert (probably after a retry). I've told it to just log those and proceed, and I can get it to pass on my machine. It takes about 10 minutes.

adamnovak commented:
OK, I think this is ready on my end.

hannes-ucsc commented on src/toil/jobStores/azureJobStore.py:
```diff
@@ -135,7 +148,13 @@ def create(self, command, memory, cores, disk,
+            # A conflict on insert probably means we retried but already
+            # succeeded. Just ignore it. We made a _newJobID() ourselves, so
+            # unless that's broken we can't be re-using a job ID.
+            logger.warning("Conflict when inserting job %s" % jobStoreId)
```
Although useful, this is unrelated to the issue that this PR is solving. It seems to partially address #579, partially because this here isn't the only call of insert_entity.

hannes-ucsc commented:
I'm inclined to merge this without 5b33d1cbf2d452b3bb52545ee34bb3152b0be5c6 (the commit that partially resolves #579) even though that commit is required to make `testManyJobs` pass. I think there may be a more general solution to #579.

The 3.1.x branch and master are virtually identical at this point so I don't think I need to redo the PR on that branch.

hannes-ucsc commented:
Continued in #628.

adamnovak commented:
It definitely might make sense to take care of insert_entity all at once
(maybe in the table wrapper class?).

On Fri, Dec 18, 2015 at 7:14 PM, Hannes Schmidt <notifications@github.com>
wrote:

> I'm inclined to merge this without 5b33d1c
> <https://github.com/BD2KGenomics/toil/commit/5b33d1cbf2d452b3bb52545ee34bb3152b0be5c6>
> (the commit that partially resolves #579
> <https://github.com/BD2KGenomics/toil/issues/579>) even though that
> commit is required to make testManyJobs pass. I think there may be a more
> general solution to #579 <https://github.com/BD2KGenomics/toil/issues/579>
> .
>
> The 3.1.x branch and master are virtually identical at this point so I
> don't think I need to redo the PR on that branch.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BD2KGenomics/toil/pull/585#issuecomment-165939942>.
>
